Added CSV and XLS flags to pulse cards to the Pulse API

senior opened the pull request:
Added include_csv and include_xls columns to the pulse_card table. the
REST endpoints now accept the new parameters and CSV/XLS attachments
will be included in pulse and alert emails

rcronin commented:
I'm confused. Is this supposed to make existing pulse/alerts include an attachment?

senior commented:
@rcronin This PR does not include the UI portion of the feature, but eventually there will be an option to include CSVs or XLSes on the create/update pulse and alert pages. You will be able to add it to existing pulses and new pulses.

rcronin commented:
Ah great. Is there an easy way I can hard the boolean value so it always attaches?

senior commented:
@rcronin We're not planning to add it by default as these attachments could potentially be large, we're wanting to keep it as an "opt-in" feature. Is it the case that you have a lot of pulses and would like to add it to all of them, or are you thinking you'll forget to click the add CSV button/box?

rcronin commented:
I was thinking of just hard coding it for now until the UI is shored up.

On Mon, Dec 11, 2017 at 3:06 PM, Ryan Senior <notifications@github.com>
wrote:

> @rcronin <https://github.com/rcronin> We're not planning to add it by
> default as these attachments could potentially be large, we're wanting to
> keep it as an "opt-in" feature. Is it the case that you have a lot of
> pulses and would like to add it to all of them, or are you thinking you'll
> forget to click the add CSV button/box?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metabase/metabase/pull/6556#issuecomment-350842200>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALKw9B_l_kp7lIs4P-et5vey4c3y2g03ks5s_YtEgaJpZM4Q9v-R>
> .
>

senior commented:
@rcronin 👍 If you want to manually test it out, you can update the `pulse_card` table. There's two new columns, `include_csv` and `include_xls`. You can flip those to true and the associated pulse (`pulse_card.pulse_id` -> `pulse.id`) will include the attachments.

Note that this is a PR though and it's very possible that there could be changes before it gets merged/released.

tlrobinson commented:
@metabase/core-developers  I believe this is ready for review.

mazameli commented:
Couple things:

- One of my main assumptions for this feature is that it would let me include raw data saved questions in pulses, but it looks like that's not possible in the branch.

**Expected:**

![screen shot 2018-01-15 at 10 08 33 am](https://user-images.githubusercontent.com/2223916/34956024-1582d94c-f9dc-11e7-8668-9508b9a5ec04.png)

**Actual:**

![screen shot 2018-01-15 at 10 07 36 am](https://user-images.githubusercontent.com/2223916/34956032-1c415ac4-f9dc-11e7-97ea-3d1e8e5e08b3.png)

- I didn't do a good job of specifying this, but I think the current behavior where you have to turn on email delivery before the file attachment button appears is clunky. Actually, I think the simplest solution would be to have email delivery turned _on_ by default for new pulses.

- Looks like if a Pulse hasn't been saved yet, file attachments aren't included when sending a test pulse. I think they should be; part of the reason I'd want to send a test pulse is to see what exactly I'm sending to my teammates.

- Paperclip button should have a tooltip. When the attachment is turned on, the tooltip text should change to "Remove attachment."

![screen shot 2018-01-15 at 10 13 23 am](https://user-images.githubusercontent.com/2223916/34956362-896cfd00-f9dd-11e7-8732-9bdacb7e697c.png)

- I think we'll want the paperclip and X button icons to be a little larger. I'll tweak that myself.

- I still need to test if the files actually get attached

- Here's the path for the paperclip icon. Its `fill-rule` is `nonzero`.

```
M22.162 8.704c.029 8.782-.038 14.123-.194 15.926-.184 2.114-2.922 4.322-5.9 4.322-3.06 0-5.542-1.98-5.836-4.376-.294-2.392-.195-14.266.01-18.699.077-1.661 1.422-2.83 3.548-2.83 2.067 0 3.488 1.335 3.594 3.164.06 1.052.074 3.49.053 7.107-.006.928-.013 1.891-.023 3.072l-.023 2.527c-.006.824-.01 1.358-.01 1.718 0 1.547-.39 2.011-1.475 2.011-.804 0-1.202-.522-1.202-1.38V8.699a1.524 1.524 0 0 0-3.048 0v12.567c0 2.389 1.554 4.428 4.25 4.428 2.897 0 4.523-1.934 4.523-5.06 0-.348.003-.875.01-1.691l.022-2.526c.01-1.184.018-2.15.024-3.082.021-3.697.008-6.155-.058-7.3C20.227 2.592 17.469 0 13.79 0c-3.695 0-6.438 2.382-6.593 5.737-.213 4.613-.312 16.585.01 19.21C7.697 28.94 11.53 32 16.067 32c4.482 0 8.61-3.327 8.937-7.106.168-1.935.235-7.302.206-16.2a1.524 1.524 0 0 0-3.048.01z
```

tlrobinson commented:
> One of my main assumptions for this feature is that it would let me include raw data saved questions in pulses, but it looks like that's not possible in the branch.

I was just going off of the Sketch files I found. Is there another document I missed outlining this feature?

mazameli commented:
@tlrobinson I see what happened: we've been gradually migrating Sketch files from Dropbox to Google Drive, and I've normally been good about making sure a design only exists in one place or the other. In this case, I forgot to remove the outdated file attachments design from Dropbox. The up-to-date one includes designs for raw data. Totally my fault — we should've checked in with each other to make sure you were looking at the right file. I'll send you the correct link.

tlrobinson commented:
@metabase/core-developers @mazameli What criteria should we be using for the "This question will be added as a file attachment" scenario?

Currently it looks like we don't let you select cards raw data GUI cards, so that's easy to support. However, currently any SQL card is allowed, but we show it as a table with only the first 3 columns if there are more than 3.

We could change it so > 3 columns results in "This question will be added as a file attachment", but it might break people's pulses if they were relying on this behavior.

¯\\_(ツ)_/¯

tlrobinson commented:
It's also a bit confusing only supporting attachments in email and not Slack. If you add a raw data card then switch from email to Slack what's supposed to happen?

mazameli commented:
@tlrobinson, progress! Replies to your questions, and a few more observations:

- Good question about the SQL cards. The way you chose to handle it seems to be no worse than the current UX in 0.27. That being said, we're going to want to revisit this logic because it seems to make it hard to use a big unaggregated table (like a data cube) to create a chart and include that chart in a pulse — i.e., if you don't take the pains to modify that cube down to 3 or fewer columns, we won't display your chart in the pulse. Again, this is not a regression; just something I'm realizing. Maybe we look at this in 0.28.1?
- Yeah, I feel you that it's at least slightly funky that the file attachments aren't included in Slack, but I think there are enough questions with the UX and implementation that personally I'm okay with starting with the very clear usecase of email attachments. The updated UX around this in this branch feels right enough to me.

- These two warning messages in this state seem wrong to me. I think inside the card we could just have the title of the question and then the subtext could be a lighter gray rather than yellow, saying, "This question won't be included in your Pulse." And the text on the side could say, "Heads up \n Raw data questions can only be included as email attachments"

<img width="866" alt="screen shot 2018-01-16 at 12 48 37 pm" src="https://user-images.githubusercontent.com/2223916/35011609-a6368a9e-fabb-11e7-81cf-e60ccd08e417.png">

- In the email itself, this is what I'm seeing for raw data questions now:

<img width="1004" alt="screen shot 2018-01-16 at 12 37 09 pm" src="https://user-images.githubusercontent.com/2223916/35011226-47646456-faba-11e7-8c67-7864208fbea7.png">

- Instead, we want to display this in emails for raw data questions:

<img width="321" alt="screen shot 2018-01-16 at 12 38 36 pm" src="https://user-images.githubusercontent.com/2223916/35011239-4f5d27b0-faba-11e7-903c-29dc69763ab5.png">

senior commented:
@tlrobinson @mazameli Found the problem with the attachments not being included in the test messages and just pushed up a fix.

senior commented:
@tlrobinson Pulse unit tests are failing https://circleci.com/gh/metabase/metabase/15405#tests/containers/5. Was failing before and after my pushed commit.

tlrobinson commented:
@mazameli Can you get me a png of the paperclip at the right size? Probably same size as `resources/frontend_client/app/assets/img/pulse_no_results@2x.png`. Feel free to commit it to that directory on this brach.

attekei reviewed (commented):
I reviewed the frontend code and generally it looks good, 

### Issues related to both frontend and backend
* Alerts don't seem to support attachments yet – there's no CSV/XLS option in alert creation/editing form and the alert email for a raw data query looks weird:
![screen shot 2018-01-18 at 17 00 34](https://user-images.githubusercontent.com/6034928/35104363-457a4bde-fc71-11e7-9b94-9a6a90f3f4a1.png)

### Backend-related issues
* With an empty result, currently the CSV/XLS attachment contains only the row with column names. Should the attachment be excluded completely if the result is empty?
*  If you have enabled a remapping for some table column, then the remapped values appears as a separate column in addition to the original column.  Should the original column be excluded from the CSV/XLS in the similar way as the frontend excludes it? Note that this is not a new issue but something that applies to `Download full results` feature of questions as well.
* I don't know if this was being addressed already, but I frequently get this backend error when trying to render raw data cards:
```
java.lang.IllegalArgumentException: No method in multimethod 'make-image-bundle' for dispatch value: [:attachment nil]
        at clojure.lang.MultiFn.getFn(MultiFn.java:156)
        at clojure.lang.MultiFn.invoke(MultiFn.java:233)
        at metabase.pulse.render$fn__27711.invokeStatic(render.clj:476)
        at metabase.pulse.render$fn__27711.invoke(render.clj:475)
```

### Frontend-related issues:
* It’s terrible that we don’t yet have any integration tests for pulses. Because of that I personally find modifying pulse-related frontend code pretty scary. Covering some basic interaction paths – like adding a new pulse, adding a question to it, adding an email recipient, choosing an XLS attachment and saving the pulse – would be a nice start. I don’t know if the lack of tests should be a hard blocker for merging this PR or not.

salsakran commented on frontend/src/metabase/pulse/components/PulseCardPreview.jsx:
```diff
@@ -14,22 +18,83 @@ export default class PulseCardPreview extends Component {
 
+    componentWillReceiveProps(nextProps) {
+        // if we couldn't detect a card type, set include_csv = true
+        if (nextProps.cardPreview && nextProps.cardPreview.pulse_card_type == null &&
```
👍

salsakran commented:
I'm ok with the remapped column appearing, as well as an empty table being included if it's an empty result. 

I would however like a frontend test covering at least the things in this PR (eg, CSV/XLS attachments), which I imagine would also test creating/editing the pulse.

tlrobinson commented on frontend/src/metabase/pulse/components/PulseCardPreview.jsx:
```diff
@@ -39,3 +104,58 @@ export default class PulseCardPreview extends Component {
     }
 }
+
+// copied from metabase/pulse/render.clj
```
It's the React reimplementation of the same layout. We want to show something different in the card picker than is actually rendered by the backend.

tlrobinson commented:
@senior can you handle this:

> With an empty result, currently the CSV/XLS attachment contains only the row with column names. Should the attachment be excluded completely if the result is empty?

senior commented:
@tlrobinson @salsakran It's easy enough to exclude that attachment if the resultset is empty, I could go either way on that one. I'll put a null check around the attachment stuff to get that make-image-bundle error go away

senior commented:
@attekei I think there must be something odd about your setup. The stacktrace you posted for the make-image-bundle error is that it can't find the attachment PNG that we include in the resources directory (frontend_client/app/assets/img/attachment@2x.png). How were you running metabase when you saw that error?

senior commented:
I think we need to leave the attachment in place, even though it has no rows. This is what the email looks like now:
![no results attachment](https://user-images.githubusercontent.com/215910/35125518-8c61fa1c-fc6f-11e7-8953-d34be6991917.png)
Otherwise we need to have a different email that says there's no attachment, which doesn't seem worth the effort to me.

tlrobinson commented:
@senior I fixed the missing image after @attekei posted, so that's probably why. Sorry, should have mentioned that.

tlrobinson commented:
@senior If we change the precedence in `detect-pulse-card-type` so `:empty` is higher than `nil` we'd show the "no results" screen instead of "this has been included as an attachment". Maybe that would be better.

senior commented:
@tlrobinson Just made the change you suggested and added in my code to check for results before attaching

kdoh commented on frontend/src/metabase/pulse/components/PulseCardPreview.jsx:
```diff
@@ -14,22 +18,83 @@ export default class PulseCardPreview extends Component {
-                <div className="bordered rounded flex-full scroll-x" style={{ display: !cardPreview && "none" }} dangerouslySetInnerHTML={{__html: cardPreview && cardPreview.pulse_card_html}} />
+                <div className="absolute top right p2 text-grey-2">
+                    { attachmentsEnabled && !isAttachmentOnly &&
+                        <Tooltip tooltip={hasAttachment ? "Remove attachment" : "Attach file with results"}>
```
Probably want to tag these strings with `t`

kdoh commented on frontend/test/pulse/pulse.integ.spec.js:
```diff
@@ -0,0 +1,161 @@
+            .setDisplayName("table")
+    )
+
+    for (const pulse of await PulseApi.list()) {
```
Might be nice to leave a note here about why we're deleting all pulses before this test runs.

kdoh reviewed (dismissed):
Looks pretty good. Left one note about translating strings that we should do before merging this.

mazameli commented:
I just pulled the latest (e5bc64d), and in the emails I'm still seeing the wrong thing for raw data cards:

![screen shot 2018-01-23 at 1 04 16 pm](https://user-images.githubusercontent.com/2223916/35300723-2f1a8ff4-003e-11e8-9232-c49b423e0d0d.png)

I should be seeing this:

<img width="321" alt="35011239-4f5d27b0-faba-11e7-903c-29dc69763ab5" src="https://user-images.githubusercontent.com/2223916/35300792-6fe6fdb0-003e-11e8-9885-68688125c34b.png">

tlrobinson commented:
@mazameli e5bc64d isn't the latest, 107c361 is (b755877 at the time of your comment)

It seems to be working for me:

<img width="597" alt="screenshot 2018-01-24 13 58 25" src="https://user-images.githubusercontent.com/18193/35318620-ae1b42dc-010e-11e8-8933-eae9268c4ca3.png">

Also if you can tell me the font parameters (size, weight, color) for the text in those messages I can update them.

salsakran commented:
the couple things I've tried look like they're working. @mazameli did you have any other concerns or can I merge?

salsakran commented:
@senior does this attach raw data questions to alerts?

senior commented:
@salsakran The backend supports it, I just pushed up a test to verify it works: 155868fa37c325da446453c808d7d10c88c0e80f. I think the frontend just needs to include that csv/xls selector widget from pulses. @tlrobinson is that something that is easy to get included?

salsakran commented:
I think the idea for alerts is that we'd include the csv for raw data results automatically. Otherwise the alert doesn't make a ton of sense. Right @mazameli ?

mazameli commented:
@salsakran @senior yes, that's correct.

senior commented:
@tlrobinson There's a linter failures in your latest commit

tlrobinson commented:
@senior fixed.

senior commented:
Are we good to go on this? If so I'll squash the commits and force push for merge

salsakran commented:
yup, let's merge this and make the change that adds the default behaviour of attaching raw data results alerts. If the latter is simple enough we can do it in this pr.

senior commented:
@salsakran Already in this PR, @tlrobinson is way ahead of us https://github.com/metabase/metabase/pull/6556/commits/e41a41ead406295fffca30cf5021e2a268e0e13c

iethree committed:
Add CSV/XLS Pulse attachment support to the backend

This commit adds the backend pieces needed to attach CSV and XLS files
of query results to Pulses. Attachment support is also there for
Alerts. This is an opt-in feature for pulses and alerts. Users can
update existing Pulses to include attachments or add them to a new
pulse. The file(s) will only be attached when there are results.

Fixes #6587, fixes #3894, fixes #2623

iethree committed:
Implement CSV/XLS attachment pulse UI

This commit adds attachment support to the pulse frontend. This commit
also automatically adds CSV attachments to alerts that are on "rows"
queries.
